File: plugkit/js/stream_reader.js
class StreamReader {
  // Construct a new StreamReader instance
  // @return StreamReader
  constructor() {
    this._fields = {
      length: 0,
      slices: []
    }
  }

  // Total length of the added payloads
  // @property Integer
  get length() {
    return this._fields.length
  }

  /// Add all slices in `payload` to the stream.
  addPayload(payload) {
    if (typeof payload !== 'object' || payload === null ||
        payload.constructor.name !== 'Payload') {
      throw new TypeError('First argument must be an Uint8Array')
    }
    for (const slice of payload.slices) {
      this.addSlice(slice)
    }
  }

  /// Add `slice` to the stream.
  addSlice(slice) {
    if (!(slice instanceof Uint8Array)) {
      throw new TypeError('First argument must be an Uint8Array')
    }
    this._fields.length += slice.length
    this._fields.slices.push(slice)
  }

  /// Find the first occurrence of `pattern` in the stream.
  /// The first `offset` bytes in the stream are skipped.
  ///
  /// Return the _end_ position of the found byte sequence,
  /// or -1 if no such byte sequence is found.
  // @return Integer
  search(pattern, offset = 0) {
    if (!(pattern instanceof Uint8Array)) {
      throw new TypeError('First argument must be an Uint8Array')
    }
    if (!Number.isInteger(offset)) {
      throw new TypeError('Second argument must be an integer')
    }

    if (pattern.length === 0) {
      return -1
    }

    const slices = this._fields.slices
    let beginOffset = 0
    let begin = 0
    for (; begin < slices.length && (beginOffset += slices[begin].length) <= offset; ++begin);
    if (beginOffset <= offset) {
      return -1
    }
    beginOffset -= slices[begin].length
    let front = beginOffset

    for (let i = begin; i < slices.length; ++i) {
      const slice = slices[i]
      let index = 0
      if (i === begin) {
        index = offset - beginOffset
      }
      for (; index < slice.length; ++index) {
        if (slice[index] === pattern[0]) {
          const window = this.read(pattern.length, front + index)
          if (window.length === pattern.length) {
            let equal = true
            for (let j = 0; j < window.length; ++j) {
              if (window[j] !== pattern[j]) {
                equal = false
                break
              }
            }
            if (equal) {
              return front + index + pattern.length
            }
          }
        }
      }
    }

    return -1
  }

  /// Read a byte sequence from the stream up to `length` bytes.
  /// The first `offset` bytes in the stream are skipped.
  /// @return Uint8Array
  read(length, offset = 0) {
    if (!Number.isInteger(length)) {
      throw new TypeError('First argument must be an integer')
    }
    if (!Number.isInteger(offset)) {
      throw new TypeError('Second argument must be an integer')
    }

    const slices = this._fields.slices
    let beginOffset = 0
    let begin = 0
    for (; begin < slices.length && (beginOffset += slices[begin].length) <= offset; ++begin);
    if (beginOffset <= offset) {
      return new Uint8Array()
    }
    beginOffset -= slices[begin].length
    let endOffset = beginOffset
    let end = begin
    for (; end < slices.length && (endOffset += slices[end].length) < offset + length; ++end);
    let continuous = true
    const buflen = Math.min(length, endOffset - beginOffset)
    const sliceOffset = offset - beginOffset
    if (slices[begin].length >= sliceOffset + buflen) {
      return slices[begin].slice(sliceOffset, sliceOffset + buflen)
    }
    const data = new Uint8Array(buflen)
    let dst = 0
    for (let i = begin; i <= end; ++i) {
      let slice = slices[i]
      if (i === begin) {
        slice = slice.slice(offset - beginOffset)
      }
      data.set(slice.slice(0, data.length - dst), dst)
      dst += slice.length
    }
    return data
  }
}

exports.StreamReader = StreamReader

File: plugkit/src/stream_reader.cpp
#include "stream_reader.h"
#include "payload.hpp"
#include <algorithm>
#include <cstring>
#include <vector>

namespace plugkit {

struct StreamReader {
  std::vector<Slice> slices;
  size_t length = 0;
};

StreamReader *StreamReader_create() { return new StreamReader(); }

void StreamReader_destroy(StreamReader *reader) { delete reader; }

size_t StreamReader_length(const StreamReader *reader) {
  return reader->length;
}

void StreamReader_addSlice(StreamReader *reader, Slice slice) {
  reader->slices.push_back(slice);
  reader->length += Slice_length(slice);
}

void StreamReader_addPayload(StreamReader *reader, const Payload *payload) {
  for (const Slice &slice : payload->slices()) {
    StreamReader_addSlice(reader, slice);
  }
}

size_t StreamReader_search(StreamReader *reader,
                           const char *data,
                           size_t length,
                           size_t offset) {
  if (reader->slices.empty() || length == 0) {
    return RANGE_NPOS;
  }
  size_t beginOffset = 0;
  size_t begin = 0;
  for (; begin < reader->slices.size() &&
         (beginOffset += Slice_length(reader->slices[begin])) <= offset;
       ++begin)
    ;
  size_t beginLen = Slice_length(reader->slices[begin]);
  beginOffset -= beginLen;
  size_t front = beginOffset;

  std::unique_ptr<char[]> buf(new char[length]);

  for (size_t i = begin; i < reader->slices.size(); ++i) {
    const Slice &slice = reader->slices[i];
    size_t sliceLen = Slice_length(slice);
    size_t index = 0;
    if (i == begin) {
      index = offset - beginOffset;
    }
    for (; index < sliceLen; ++index) {
      if (slice.begin[index] == data[0]) {
        const Slice &window =
            StreamReader_read(reader, &buf[0], length, front + index);
        if (Slice_length(window) == length &&
            std::memcmp(window.begin, data, length) == 0) {
          return front + index + length;
        }
      }
    }
    front += sliceLen;
  }
  return RANGE_NPOS;
}

Slice StreamReader_read(StreamReader *reader,
                        char *buffer,
                        size_t length,
                        size_t offset) {
  if (reader->slices.empty()) {
    return Slice{buffer, buffer};
  }
  size_t beginOffset = 0;
  size_t begin = 0;
  for (; begin < reader->slices.size() &&
         (beginOffset += Slice_length(reader->slices[begin])) <= offset;
       ++begin) {
  };
  if (beginOffset <= offset) {
    return Slice{buffer, buffer};
  }
  beginOffset -= Slice_length(reader->slices[begin]);
  size_t endOffset = beginOffset;
  size_t end = begin;
  for (; end < reader->slices.size() &&
         (endOffset += Slice_length(reader->slices[end])) < offset + length;
       ++end)
    ;
  bool continuous = true;
  for (size_t i = begin; i < end; ++i) {
    if (reader->slices[i].end != reader->slices[i + 1].begin) {
      if (!buffer) {
        const char *data = reader->slices[begin].begin + (offset - beginOffset);
        return Slice{data, reader->slices[i].end};
      }
      continuous = false;
      break;
    }
  }
  size_t buflen = std::min(length, endOffset - beginOffset);
  if (continuous) {
    const char *data = reader->slices[begin].begin + (offset - beginOffset);
    return Slice{data, data + buflen};
  }
  char *dst = buffer;
  for (size_t i = begin; i <= end; ++i) {
    Slice slice = reader->slices[i];
    if (i == begin) {
      slice.begin += (offset - beginOffset);
    }
    size_t capacity = buffer - dst;
    size_t sliceLen = std::min(Slice_length(slice), capacity);
    std::memcpy(dst, slice.begin, sliceLen);
    dst += sliceLen;
  }
  return Slice{buffer, buffer + buflen};
}
} // namespace plugkit
